### DISCUSSION TITLE
Where do I have to call render() for clientside components?

### DISCUSSION THREAD

**Discussion by jonasfroeller** [honojs/hono#2609] [Q&A]
I am using deno and I had problems importing honos jsx-runtime.  
Setting it in deno.json caused `error: pragma cannot be set when runtime is automatic`, because I have to set `/** @jsx jsx */`, so that my editor stops screaming at me, that the jsx I write is invalid.

deno.json:
```json
"compilerOptions": {
  "jsx": "precompile",
  "jsxImportSource": "hono/jsx/dom"
},
"imports": {
  "hono/jsx/dom/jsx-runtime": "https://deno.land/x/hono@v4.3.1/jsx/jsx-runtime.ts"
}
```
Error: Cannot find name 'React'.deno-ts(2304)

I now did it with `import { jsx as _jsx } from "https://deno.land/x/hono@v4.3.1/jsx/dom/jsx-runtime.ts"; `. I am not sure if this works properly, because the error `Uncaught ReferenceError: document is not defined` says, it doesn't, right?  

Now to the actual question: Where am I supposed to call render? If I put it at the end of the code I get error: `Uncaught ReferenceError: document is not defined`, because the code is serverside, right? So where do I have to put the render call?
```ts
const root = document.getElementById('root') as HTMLElement;
render(<Chat />, root);
```

I dont really know what to do, the documentation is kind of confusing to me.

**Comment by nakasyou** [comment]
`hono/jsx/dom` is exists for working on client side.

You may run this code on Deno. Deno is local runtime, so `document` is not exists. Rendering target is not exists too.

If you want to Hono JSX on server-side rendering, you should use [JSX helper](https://hono.dev/guides/jsx#jsx).

If you want to Hono JSX for frontend with Deno, you have to convert TypeScript code to frontend JavaScript code using compiler or framework such as [esbuild](https://esbuild.github.io), [Fresh](https://fresh.deno.dev), [Vite](http://vite.dev) or other.

Ask me about that if your code can't work well.

**Comment by jonasfroeller** [comment]
@nakasyou [JSX helper](https://hono.dev/guides/jsx#jsx) is what I tried. That is the problem :).  
I do not want to include another framework, because it is only needed for one route.
